Controllers/UserController.js

```diff
@@ -1,21 +1,91 @@
-import User from '../Models/User.js'
+import User from '../Models/User.js';
+import StandUpHelpers from '../utils/StandUpHelpers.js';
+
 import { validationResult } from 'express-validator';
 
 export default {
-  
-    users: (req, res) => {
-        User.find().then(users => {
-            res.json({users:users});
+  users: (req, res) => {
+    User.find().then((users) => {
+      res.json({ users: users });
+    });
+  },
+
+  updateUser: async (req, res) => {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+      return res.status(400).json({ errors: errors.array() });
+    }
+
+    try {
+      var user = await User.findOne({
+        username: req.user.username,
+      })
+        .select("standups configs.timeZone")
+        .lean();
+
+      if (!user) throw new Error(`User is not found.`);
+      if (
+        req.body.user.configs.timeZone != user.configs.timeZone &&
+        user.standups.length != 0
+      ) {
+        user.configs.timeZone = req.body.user.configs.timeZone;
+        StandUpHelpers.updateStandUpRemindersByUser(
+          user,
+          user.standups,
+          true,
+          true,
+          true
+        );
+      }
+
+      user = await User.findByIdAndUpdate(
+        user._id,
+        {
+          $set: { configs: req.body.user.configs },
+        },
+        { new: true }
+      );
+      return res.json({ sucess: true, user: user })
+    } catch (e) {
+      return res
+        .status(404)
+        .json({ message: "Error updating user", error: e.message });
+    }
+  },
+
+  deleteUser: async function (req, res) {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+      return res.status(400).json({ errors: errors.array() });
+    }
+
+    try {
+        var user = await User.findOne({
+          username: req.user.username,
         })
-      },
+          .select("standups configs.timeZone")
+          .lean();
+  
+      if (user.standups.length != 0)
+        await StandUpHelpers.updateStandUpRemindersByUser(
```

Comment:
What does this do?

Reply:
## What does it do
This does a simple check to see if the about to delete user is subscribed to any standups. Afterward, it will move on to check if each of the standup the user is subscribed to has a reminder.
If the standup has reminders, it will proceed to remove as described:

When a user subscribes to a standup that has reminders, they will be added to that stack of reminders.
For instance, this is what a snippet of one reminder looks like:
- with StaticTime:
```json
{
    "time": {
        "hour": 10,
        "min": 20
    },
    "_id": "623b7f05fa8eb8e4dd2d665c",
    "list": [
        {
            "user_id": "6243286395fe08725ab2082f",
            "notification_time": "2022-03-30T02:20:00.000Z"
        },
        {
            "user_id": "623b0d129500e3c0c6c42617",
            "notification_time": "2022-03-30T15:20:00.000Z"
        }
    ]
}
```
- with no staticTime:
```json
{
    "time": {
        "hour": 10,
        "min": 20
    },
    "_id": "6242af4782be4558db1282ec",
    "list": [
        {
            "user_id": [
                "623b0d129500e3c0c6c42617",
                "6243286395fe08725ab2082f"
            ],
            "notification_time": "2022-03-30T10:20:00.000Z"
        }
    ]
}
```

Therefore, in the case, if the user with id "623b0d129500e3c0c6c42617" is deleted the corresponding reminders that have this particular user in the stack should also be removed.
```
Which means:
- reminders that enable staticTime, the object that has the user_id corresponding to the deleted user will be pulled out
- reminders without staticTime, user will be pulled out from the list of user_id
```

## The reason to have it in a helper file
The reason why I moved the functionality to a helper file is that I see there are multiple occurrences of making the same queries depending on the user's actions. Therefore, I did this to minimize repeating written lines of queries.
The cases:
- if a user attempts to subscribe to standup with reminders, they will need to be added to the corresponding reminder stack. `(update)`
- if a user changes their time zone, and if the standups that the user subscribes to that have staticTime enabled it will also need to be updated according to the new time zone that the user set. This is done by removing the existing reminder from the stack and regenerating it again. `(remove, update)`
  - This is because reminders in this case are generated depending on user's time zone.
- if a user unsubscribes from a standp with reminders, they should be removed from the corresponding reminder stack. `(remove)`
- if a user deletes their account, the dependents will also need to be removed. In the current case it will be standup reminders that has the to-be deleted user in the stack. `(remove)`

Reply:
If I am understanding the second point stated right;
- it means that upon the author schedules a reminder, only the `notification_time` field will be generated. The cron job checks based on the `notification_time` and only if a reminder is ready to be sent only then it will query the list of users to send the reminders to.
- Therefore, there is no need to include the user's particular (id) inside the reminders property of standups.

If the idea is as stated above,
1. This would work for reminders without staticTime enabled
2. However, in the case of standups with staticTime enabled it would be more of a necessity:
- As different users would have different reminder times scheduled, cause this option is to have reminders sent out conformed to each user's timezone
- This leads to having a need to keep track of which schedule time is belonging to which user

How do you see this @Fenn-CS?

Reply:
Does it mean that instead of keeping track of the timezone we should store the offset instead?

## Reason to store/keep track of user timezone
Currently, in the case of staticTime, the `notification_time` is dependent on the user timezone set. Here are some of the cases:

1. When users subscribes to the standup with staticTime on for reminders:
- First, generate a DateTime that is in the user timezone
- Afterwards, convert that generated DateTime into UTC to be stored
- Then the user schedule gets added to the list as shown in the snippet below for staticTime
2.  A user decides to change their timezone
- When the user makes the change, it will check through all the standups that the user is associated with to see if any of the standup has staticTime enabled
- If there are any standups with staticTime enabled, subsequent updates will be made to regenerate user reminder time again according to the new timezone
3. If the standup author decides to enable staticTime option
- In this case, if there are existing users subscribe to this standup it will loop through the list of user and generate the corresponding reminders according to their timezone first, which will then be converted into UTC to store in the database
```
Therefore, the flow to generate a user reminder for staticTime is always:
- (1) Generate a DateTime in user timezone → (2) convert that DateTime into UTC for storing
However, in any case if anyone of the 3 cases stated happens there is a need to access user timezone again to get the updated reminder datetime
```
There is another case maybe if we decide to implement showcasing the list of reminders that the users have. This will require access to the user timezone to convert UTC back to display the datetimes.

## Expensive looping to send reminders for each standup schedule (staticTime)
The current approach is that:
1. Cron job runs at every 3-minute interval to query which standup has reminders ready to be sent
- This will return a list of standups
- The standup collection has a index based on the notification_time
2. Afterwards, it runs a map method against the list of standups. It will first check if today is a day to send out reminders
3. There will be another inner map function running against each `schedules` for the standup after passing checks mentioned in `2`
- In the case of staticTime, for each schedule it will check for each user scheduled time in the list to see if they are ready to be send out or not `(user.notification_time < date)`
- Probably this is the taxing part? as it has to check through the list of all users in the schedule list
4. After passing the check from `3`, it waits on the reminders to be sent out which will then proceed to make the update for the next schedule
- Here it would not wait for the update to be resolved, instead it will return a promise and it moves on to the next user
5. All these accumulated nested promises are wrapped in Promise.all to be run paralelly for this current batch

If the expensive looping is referring to `3` I have not come across a direct method for the document to return only those specific `notification_time < the current datetime`. I have only seen reference answer pointing to do [aggregation](https://www.mongodb.com/docs/manual/reference/operator/aggregation-pipeline/#std-label-aggregation-pipeline-operator-reference) on the fields. If this is the case Ill see to it how it can be done

> code to loop through the list of standups with reminders
```js
 const batchProcess = standups.map(async (standup) => {
      if (standup.reminders.days.includes(date.weekday)) {
        return standup.reminders.schedules.map(async (schedule) => {
          if (standup.reminders.staticTime) {
            return schedule.list.map(async (user) => {
              if (user.notification_time < date) {
                try {
                  await this.sendReminders(user.user_id, standup);

                  const nextReminder = this.nextReminder(
                    user.notification_time,
                    'utc',
                  );

                  return StandUp.findOneAndUpdate(
                    { _id: standup._id },
                    {
                      $set: {
                        'reminders.schedules.$[schedule].list.$[user].notification_time':
                            nextReminder,
                      },
                    },
                    {
                      arrayFilters: [
                        { 'schedule._id': schedule._id },
                        { 'user.user_id': user.user_id },
                      ],
                    },
                  ).exec();
                } catch (e) {
                  console.error('Error while sending reminders: ', e.message);
                }
              }
            });
          }
          try {
            if (schedule.list[0].notification_time < date) {
              await Promise.all(
                schedule.list[0].user_id.map(async (user) => this.sendReminders(user, standup)),
              );

              const nextReminder = this.nextReminder(
                schedule.list[0].notification_time,
                'utc',
              );

              return StandUp.findOneAndUpdate(
                { _id: standup._id },
                {
                  $set: {
                    'reminders.schedules.$[schedule].list.$[].notification_time':
                        nextReminder,
                  },
                },
                { arrayFilters: [{ 'schedule._id': schedule._id }] },
              ).exec();
            }
          } catch (e) {
            console.error('Error while sending reminders: ', e.message);
          }
        });
      }
    });
    
    await Promise.all(batchProcess);
```

> example of a staticTime enabled reminder snippet in a standup
```json
"reminders": {
    "staticTime": true,
    "days": [1, 2, 3, 4, 5],
    "schedules": [
        {
            "time": { "hour": 10, "min": 20 }, //This is one reminder
            "_id": "6244031c407effaf82a5cc3e",
            "list": [
                 {
                     "user_id": "6244023e407effaf82a5cc32",
                     "notification_time": "2022-03-31T02:20:00.000Z"
                 },
                 {
                      "user_id": "624402f8407effaf82a5cc38",
                      "notification_time": "2022-03-31T15:20:00.000Z"
                  }
              ]
          },
          {
              "time": {  "hour": 14, "min": 0 }, //This is another reminder
              "_id": "6244031c407effaf82a5cc3f",
              "list": [
                  {
                      "user_id": "624402f8407effaf82a5cc38",
                      "notification_time": "2022-03-31T19:00:00.000Z"
                  },
                  {
                      "user_id": "6244023e407effaf82a5cc32",
                      "notification_time": "2022-03-31T06:00:00.000Z"
                  }
              ]
          }
     ]
},
```